File: coding_skill_cpp/05_state_machine.cpp
/* C++ State Machine - enum class + switch */
#include <iostream>
#include <string>
#include <map>
#include <functional>
using namespace std;

enum class State { IDLE, RUNNING, STOPPED };

class StateMachine {
    State current = State::IDLE;
    map<State, function<void()>> onEnter;
    
public:
    void setOnEnter(State s, function<void()> f) { onEnter[s] = f; }
    
    void transition(State next) {
        cout << "[SM] " << static_cast<int>(current) 
             << " → " << static_cast<int>(next) << endl;
        current = next;
        if (onEnter[current]) onEnter[current]();
    }
    
    State getState() const { return current; }
};

int main() {
    cout << "\n=== C++ State Machine ===" << endl;
    
    StateMachine sm;
    sm.setOnEnter(State::RUNNING, []() { cout << "  → 실행 시작!" << endl; });
    sm.setOnEnter(State::STOPPED, []() { cout << "  → 정지됨" << endl; });
    
    sm.transition(State::RUNNING);
    sm.transition(State::STOPPED);
    sm.transition(State::IDLE);
    
    return 0;
}

File: coding_skill_py/05_state_machine.py
"""Python State Machine - Enum + dict"""
from enum import Enum, auto
from typing import Callable, Dict

class State(Enum):
    IDLE = auto()
    RUNNING = auto()
    STOPPED = auto()

class StateMachine:
    def __init__(self):
        self.current = State.IDLE
        self.on_enter: Dict[State, Callable] = {}
    
    def set_on_enter(self, state: State, callback: Callable):
        self.on_enter[state] = callback
    
    def transition(self, next_state: State):
        print(f"[SM] {self.current.name} → {next_state.name}")
        self.current = next_state
        if next_state in self.on_enter:
            self.on_enter[next_state]()

def main():
    print("\n=== Python State Machine ===")
    
    sm = StateMachine()
    sm.set_on_enter(State.RUNNING, lambda: print("  → 실행 시작!"))
    sm.set_on_enter(State.STOPPED, lambda: print("  → 정지됨"))
    
    sm.transition(State.RUNNING)
    sm.transition(State.STOPPED)
    sm.transition(State.IDLE)

if __name__ == "__main__":
    main()
